### PR TITLE
feat: add parser for canon.lock files

### PR DESCRIPTION
**Pull request by DmitriyLewen**
## Description
Added parser for [canon.lock](https://docs.conan.io/en/latest/versioning/lockfiles/introduction.html#locking-dependencies) files

## Related issues
- aquasecurity/trivy/issues/2708

### CODE REVIEW

File: pkg/c/conan/lock/parse.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
+package lock
+
+import (
+	"encoding/json"
+	"strings"
+
+	dio "github.com/aquasecurity/go-dep-parser/pkg/io"
+	"github.com/aquasecurity/go-dep-parser/pkg/types"
+	"golang.org/x/xerrors"
+)
```
Comment (afdesk) [inline pkg/c/conan/lock/parse.go:10]: doesn't `golangci-lint` show `goimport` error?

File: pkg/c/conan/lock/parse.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	}
+
+	for _, nod := range lock.GraphLock.Nodes {
+		if nod.Ref != "" {
+			// ref format examples: 'pkga/0.1@user/testing'
+			// 'pkgb/0.1.0'
+			// 'pkgc/system'
+			ref := strings.Split(strings.Split(nod.Ref, "@")[0], "/")
+			if len(ref) != 2 {
+				return nil, nil, xerrors.Errorf("unable to parse ref: %s", nod.Ref)
```
Comment (afdesk) [inline pkg/c/conan/lock/parse.go:46]: maybe i miss something.
there will be an error even for one broken ref. right?

I'd prefer to write logs and continue. but if we don't have logging, breaking is better way.

File: pkg/c/conan/lock/parse.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+			}
+
+			// skip system dependencies
+			if ref[1] == "system" {
+				continue
+			}
```
Comment (afdesk) [inline pkg/c/conan/lock/parse.go:52]: why should we skip `system` deps?

Comment (DmitriyLewen) [reply pkg/c/conan/lock/parse.go]: conan system dependencies have a `system` only version (e.g. [glu](https://conan.io/center/glu))

At first I thought that we need to skip these dependencies, but currently I think we need to keep all dependencies.

Thank for your note! Changed it.

File: pkg/c/conan/lock/parse.go (hunk 2 above)
Comment (DmitriyLewen) [reply pkg/c/conan/lock/parse.go]: It makes sense. 
I have added a debug message. Can you take a look?

File: pkg/c/conan/lock/parse.go (hunk 1 above)
Comment (DmitriyLewen) [reply pkg/c/conan/lock/parse.go]: Fixed

File: pkg/c/conan/lock/parse.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+type LockFile struct {
+	GraphLock GraphLock `json:"graph_lock"`
+}
+
+type GraphLock struct {
+	Nodes map[string]Nod `json:"nodes"`
+}
+
+type Nod struct {
```
Comment (knqyf263) [inline pkg/c/conan/lock/parse.go:23]: Any reason we cannot use `Node`?

File: pkg/c/conan/lock/parse_test.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+		{
+			name:      "sad path. wrong ref format",
+			inputFile: "testdata/sad.lock",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			f, err := os.Open(tt.inputFile)
+			require.NoError(t, err)
```
Comment (knqyf263) [inline pkg/c/conan/lock/parse_test.go:51]:
```suggestion
			f, err := os.Open(tt.inputFile)
			require.NoError(t, err)
			defer f.Close()
```

File: pkg/c/conan/lock/parse.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+import (
+	"encoding/json"
+	"strings"
+	"sync"
+
+	dio "github.com/aquasecurity/go-dep-parser/pkg/io"
+	"github.com/aquasecurity/go-dep-parser/pkg/log"
+	"github.com/aquasecurity/go-dep-parser/pkg/types"
+
+	"golang.org/x/xerrors"
```
Comment (knqyf263) [inline pkg/c/conan/lock/parse.go:12]: nit

```suggestion
	"golang.org/x/xerrors"

	dio "github.com/aquasecurity/go-dep-parser/pkg/io"
	"github.com/aquasecurity/go-dep-parser/pkg/log"
	"github.com/aquasecurity/go-dep-parser/pkg/types"
```

File: pkg/c/conan/lock/parse_test.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
+package lock
```
[the remark this answers is not in the document]
Comment (DmitriyLewen) [reply pkg/c/conan/lock/parse_test.go]: Done

File: pkg/c/conan/lock/parse_test.go (hunk 5 above)
Comment (DmitriyLewen) [reply pkg/c/conan/lock/parse_test.go:52]: Done

File: pkg/c/conan/lock/parse.go (hunk 4 above)
Comment (DmitriyLewen) [reply pkg/c/conan/lock/parse.go]: This is my typo, fixed. Thanks.

File: pkg/c/conan/lock/parse.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+	}
+
+	for _, nod := range lock.GraphLock.Nodes {
+		if nod.Ref != "" {
+			// ref format examples: 'pkga/0.1@user/testing'
+			// 'pkgb/0.1.0'
+			// 'pkgc/system'
+			ref := strings.Split(strings.Split(nod.Ref, "@")[0], "/")
+			if len(ref) != 2 {
+				once.Do(func() {
```
[the remark this answers is not in the document]
Comment (DmitriyLewen) [reply pkg/c/conan/lock/parse.go]: You are right. User should see all wrong dependencies.
Removed this.

File: pkg/c/conan/lock/parse.go (hunk 6 above)
Comment (DmitriyLewen) [reply pkg/c/conan/lock/parse.go]: Done

**Review by knqyf263 — APPROVED** [review]

**Pull request merged**